### ISSUE TITLE
HDBSCAN testing

### ISSUE THREAD

**Issue by nsankar** [oracle/tribuo#246]
We have a synthetically generated training and (anomaly) test data (_attached herewith_) for testing Tribuo HDBSCAN . We first trained a model with the *python* HDBSCAN and we predicted the anomalies from the synthetic test/anomaly data . 

Then we used the same data to create a Tribuo HDBSCAN model following its standard workflow. (The crux of the code flow is attached herewith for ref.) and used the test data for predictions. However, Tribuo model did not predict any anomaly against the test data including when I played aound changing the  hdbscan parameters . All the predicted score was pretty low like 0.10809591 .. using the getScore() method from prediction.  My understanding from the Tribuo tutorial is that the low value predicted scores like 0.1, 0.2 etc are not anomalies and scores >=0.8 are anomalies . Is that correct?

Also, In general w.r.t Tribuo, I was able to test the Tribuo HDBSCAN model training and prediction working fine with the real world large size card holder labeled  dataset mentioned as a benchmark with Python HDBSCAN in the Tribuo HDBSCAN paper, and I was able to test its labeled anomalies working. Further , the smaller toy  dataset in the Tribuo repo is another thing that works fine.

Based on this,my question is what  might be the possible reasons that I am unable to get anomaly outputs from Tribuo from my test data Vis-a-vis Python HDBSCAN? Does Tribuo's HDBSCAN require more training data than its python counterpart? or Is it dependent on any data characteristics?

Attachments:-

train and test data
[tribuo-syn-test-data.zip](https://github.com/oracle/tribuo/files/9045506/tribuo-syn-test-data.zip)

crux of the hdbscan code flow.
[tribuo_code_crux.txt](https://github.com/oracle/tribuo/files/9045519/tribuo_code_crux.txt)

**Comment by Craigacp** [comment]
Is this with Tribuo v4.2? The way noise points are tagged at inference time has been changed in `main` as it wasn't consistently marking things as noise when they are far away from the original training points. If you've not already done this could you test it out on the current head of `main`?

**Comment by nsankar** [comment]
@Craigacp  The one I have tested was in the current head of main only and not in v4.2 There is also another observation. According to the tutorial, the predictions with higher scores like 0.8 and above are mostly outliers . Also, per Tribuo, the labeling for the outlying points are to be 0. But , on the contrary , for very low prediction scores like 0.01, 0.02 etc which means it is not an outlier, I am getting labels as 0 in the above test which should not be the case. Pls. see the attached file of predictions on the test dataset with the last 2 columns where I have appended the prediction scores and the predicted label.
[synthetic_data_runtime_stressed_resptime_5thJuly.csv](https://github.com/oracle/tribuo/files/9047407/synthetic_data_runtime_stressed_resptime_5thJuly.csv)
?

**Comment by Craigacp** [comment]
The scores are a function of which cluster exemplar the new point was assigned to or a pre-computed outlier score if it is assigned as noise, they are not a function of distance between that exemplar and the test point.

The inference implementation in Tribuo is approximate so it doesn't need to keep the whole dataset around at inference time, it computes distances to the cluster exemplars and assigns the label based on the closest exemplar. Which Python implementation are you comparing against?

@geoffreydstewart could you take a look at this prediction issue?

**Comment by nsankar** [comment]
@Craigacp noted. I am testing  some more datasets. Will report any observed discrepancies if any In this regard. Thanks.

**Comment by geoffreydstewart** [comment]
Hello, I can try the attached datasets to see if I observe any unexpected results. I'll be back from vacation next week and will look into this then.

**Comment by geoffreydstewart** [comment]
@nsankar I have performed some tests using this dataset of 2500 records and 1013 features. Before considering predictions, first we should look at the model that is trained with this data. To do this, I prepared both a Python hdbscan and Tribuo Hdbscan notebook in this github repo [here.](https://github.com/geoffreydstewart/TribuoHdbscan/tree/syn-test/notebooks/Compare%20Labels%20and%20Outlier%20Scores/Synthetic%20Data)

The Python model identifies only 2 clusters containing 5 points and 6 points respectively. The rest of the points are all outliers. The outlier scores for many of the outlier points have very low values which are not typical outlier scores for noise points. 

The Tribuo model does not identify any clusters. All the points are determined to be outliers. Again, the outlier scores are mostly very low values which are not typical outlier scores for noise points. A model with characteristics like this will not be very useful when making predictions.

In general, I am not sure that the HDBSCAN algorithm is the best choice for this particular dataset. Perhaps you could try using KMeans and experiment with different values for k to see what clusters are established. Otherwise for outlier/anomaly detection, you may want to consider using a LibSVM anomaly detection as described [here.](https://tribuo.org/learn/4.2/tutorials/anomaly-tribuo-v4.html)

**Comment by nsankar** [comment]
@geoffreydstewart  Thanks for the inputs. On one such similar data, I tuned the minclusterSize parameter using an optimizer such that the HDBSCAN optimizes (maximizes) distinct number of clusters. (like from coarse grained to fine grained distinct clusters with good boundaries). Before I did this, all the labels for my test were '0' labels indicating ots an outlier, but the scores were pretty low on a scale of 0 to 1 like you found above. However, after optimizing the minclustersize, all the scores became 0.999

**Comment by nsankar** [comment]
@geoffreydstewart For performance,  what is the  **max. safe**  no. of parallel thread setting I can keep in the HDBSCAN trainer? based on your experiences, pls. suggest. thanks in advance.

**Comment by geoffreydstewart** [comment]
@nsankar I have not experimented with extremely large values for `numThreads` when training HDBSCAN models. Generally, the most notable performance gains are achieved by increasing the number of threads up to the number of cores available on the test system. Beyond that, I haven't seen much evidence of significant gains.
Keep in mind that the only step of the algorithm being parallelized are the neighbour queries, while the rest of the algorithm is executed in a single thread. In Tribuo 4.3 we're adding a K-D Tree structure which should always out perform a brute force technique for neighbour queries. However, in practice we have observed that training HDBSCAN models on bigger machines with lots of cores, using a `NeighboursBruteForce` implementation actually performs better. This is due to the fact that data vectors are retained in the CPU cache which causes the next step of the algorithm to execute much faster. We are continuing to work on optimizations of this algorithm.
Hopefully these details assist with your optimization activities.

**Comment by nsankar** [comment]
@geoffreydstewart @Craigacp 

I was able to serialize the HDBSCAN transformed model with the below code. I didn't know how to deserialize it properly using the read object as below..

Could you pls. help me with the snippet for deserialization ? . Thanks in advance.

`String model_save_path = "C:\\Users\\Lenovo\\Documents\\tribuo\\ss\\model_serialized";
        String outputModelPath = model_save_path.replace("\\", "/");
        final Path outputModelPath = Paths.get(outputModelPath,"SerModel.bin");
       ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(outputModelPath.toFile()));
        oos.writeObject(transformedModel);`

Deserializarion ?
`ObjectInputStream mois = new ObjectInputStream(new BufferedInputStream(new FileInputStream(modelPath.toFile())))
?? = <?? > mois.readObject();`
    	





ObjectInputStream mois = new ObjectInputStream(new BufferedInputStream(new FileInputStream(modelPath.toFile())))
model = () mois.readObject();

**Comment by Craigacp** [comment]
Model serialization and deserialization is covered in the intro tutorial here - https://tribuo.org/learn/4.2/tutorials/irises-tribuo-v4.html#Loading-and-saving-models. You should load it in as a `Model<?>` then perform the appropriate `Model.validate()` check for your prediction type (in this case `ClusterID`).

**Comment by nsankar** [comment]
Thanks @Craigacp
